Template stacks: Continuous auto versioning

suskin opened the pull request:
## Overview

When we build an artifact from our continuous build job which runs
against changes to `master`, we want to be able to consistently have a
prerelease semantic version, which follows the semantic version ordering
contract. To do this, we can generate and manipulate the version which
is used for the artifact. We use the most recent tag in the current
branch as a base.

As part of this change, we also now expect the publish pipeline to
**not** generate a version; an operator must specify the parameter by
hand.

## Testing

* I've tested the different cases of tag generation locally by hand
* Here's a [test run of the parameter-checking logic for publish](https://jenkinsci.upbound.io/job/upbound/job/suskin-test/job/master/9/console)
* Here's a [test run of the tag-generation logic for the at-tag case](https://jenkinsci.upbound.io/job/crossplane/job/app-wordpress/job/continuous-publish/job/master/13/console)

suskin committed:
versioning: auto-increment continuous build versions

When we build an artifact from our continuous build job which runs
against changes to `master`, we want to be able to consistently have a
prerelease semantic version, which follows the semantic version ordering
contract. To do this, we can generate and manipulate the version which
is used for the artifact. We use the most recent tag in the current
branch as a base.

As part of this change, we also now expect the publish pipeline to
**not** generate a version; an operator must specify the parameter by
hand.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
docs: update release documentation

The main change in this changeset is to update the release documentation
to reflect the current methods, and to talk about when a release branch
may be warranted.

There are also some minor updates to use the new name for the sample
wordpress stack.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

displague commented on vars/runStackPublishPipeline.groovy:
```diff
@@ -21,20 +21,24 @@ def call() {
         stages {
+            stage('Check Required Parameters') {
+                steps {
+                    sh "test -n '${params.version}' || ( echo 'The version is required. Please specify a version to release.' >&2 && exit 1)"
```
Should we test for a leading v?

suskin replied:
I could go either way on that, but I ended up not doing it to have dumber test logic. I suppose I could always use a `test grep -q` though, which is fun

suskin replied:
Done. Test runs:

* [Bad input](https://jenkinsci.upbound.io/job/upbound/job/suskin-test/job/master/15/console)
* [Good input](https://jenkinsci.upbound.io/job/upbound/job/suskin-test/job/master/14/console)

displague commented on docs/pipelines.md:
```diff
@@ -76,6 +82,27 @@ let you specify any parameters, and the job may fail. After the first
+If the patch is for a version before the most recent release, a release
+branch will be needed.
+
+To create a release branch, use the job `branch-create`, and use the
```
We could link to the jobs, right?

suskin replied:
Not really, since each stack will have its own copy of the jobs, and this document is common to all stacks. We could probably link to the source code for the job, but it seemed less helpful.

muvaf commented on vars/runStackContinuousPipeline.groovy:
```diff
@@ -28,6 +28,7 @@ def call() {
                 steps {
                     sh 'mkdir bin'
                     sh "curl -sL https://raw.githubusercontent.com/crossplane/crossplane-cli/${CROSSPLANE_CLI_RELEASE}/bootstrap.sh | env PREFIX=${WORKSPACE} RELEASE=${CROSSPLANE_CLI_RELEASE} bash"
+                    sh "curl -sL -o bin/semver https://raw.githubusercontent.com/fsaintjacques/semver-tool/master/src/semver && chmod +x bin/semver"
```
Can we use commit tag instead of `master` so that we can be sure the executable we run is always the same?

suskin replied:
Good call

suskin replied:
Done

suskin committed:
publish: fix missing docker login

Docker login was missing from the jobs which publish docker artifacts,
and that's needed.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>
